lib/index.js

```diff
@@ -2,3 +2,163 @@
  * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
  */
 'use strict';
+const {generateId} = require('bnid');
+const {snakeCase} = require('snake-case');
+const {
+  InvalidRequest, AccessDenied
+} = require('@interop/oauth2-errors');
+
+const {ClientMetadata, FORBIDDEN_FIELDS} = require('./ClientMetadata');
+
+function handleClientRegistration({authentication, register,
+  defaults = {}, allowClientProvidedId = false, logger}) {
+  return async (req, res) => {
+    try {
+      const initialAccessToken = await _authenticate({
+        req, authentication, logger});
+
+      const {body: registrationBody} = req;
+
+      const registration = _parseRegistration({registrationBody, defaults});
+
+      const result = registration.validateNew();
+      if(!result.valid) {
+        throw result.error;
+      }
+
+      // client can be optionally overriden by the register() callback
+      await _ensureClientId({registration, allowClientProvidedId});
+      await _generateClientSecret({registration, defaults});
+
+      // registration.registration_access_token = '...';
+      // registration.registration_client_uri = (new URL('/register/' +
+      //   encodeURIComponent(registration.client_id), baseUrl)).toString();
+
+      // Epoch time
+      registration.client_id_issued_at = Math.floor(Date.now() / 1000);
+
+      let client;
+      try {
+        client = await register({
+          registration, credentials: {initialAccessToken}
+        });
+      } catch(e) {
+        if(logger) {
+          logger.error(e);
+        }
+        throw new InvalidRequest({
+          description: 'Error registering client.'
+        });
+      }
+
+      _respond({client, res});
+    } catch(error) {
+      if(logger) {
+        logger.error(error);
+      }
+      error.respond(res);
```

Comment:
is it guaranteed that the error thrown inside here will have a `respond` method?  I gather that the imported error classes will have that but what about more mundane errors?

Reply:
Right, no, not guaranteed. We can remove that part for the moment and just throw the error. (I want to come back to this issue in the next release though.)

Reply:
This is broken due to an incorrect assumption about what the error is which will result in an error like `error.respond` is not a function.  I think it's important that this is fixed before a 1.0 release.

Reply:
Refactored.